File: algorithmAndDataStructure/cpp/minStack/minStack.cpp
#include <iostream>
#include <stack>
#include <limits.h>

using namespace std;

class MinStack {
public:
    /** initialize your data structure here. */
    stack<int> values;
    int minValue;
    MinStack() {
        minValue = INT_MAX;
    }
    
    void push(int x) {
        if (x <= minValue) {
            values.push(minValue);
            minValue = x;
        }
        values.push(x);
    }
    
    void pop() {
        int value = values.top();
        values.pop();
        if (value == minValue) {
            minValue = values.top();
            values.pop();
        }
    }
    
    int top() {
        return values.top();
    }
    
    int getMin() {
        return minValue;
    }
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack* obj = new MinStack();
 * obj->push(x);
 * obj->pop();
 * int param_3 = obj->top();
 * int param_4 = obj->getMin();
 */

File: algorithmAndDataStructure/golang/minStack/minStack.go
package minStack

import "math"

type MinStack struct {
	stack    []int
	minValue int
}

/** initialize your data structure here. */
func Constructor() MinStack {
	return MinStack{
		stack:    make([]int, 0, 0),
		minValue: math.MaxInt32,
	}
}

func (this *MinStack) Push(x int) {
	if x <= this.minValue {
		this.stack = append(this.stack, this.minValue)
		this.minValue = x
	}
	this.stack = append(this.stack, x)
	return
}

func (this *MinStack) Pop() {
	top := len(this.stack) - 1
	topValue := this.stack[top]
	this.stack = this.stack[:top]
	if topValue == this.minValue {
		this.minValue = this.stack[top-1]
		this.stack = this.stack[:top-1]
	}
}

func (this *MinStack) Top() int {
	top := len(this.stack) - 1
	return this.stack[top]
}

func (this *MinStack) GetMin() int {
	return this.minValue
}

/**
 * Your MinStack object will be instantiated and called as such:
 * obj := Constructor();
 * obj.Push(x);
 * obj.Pop();
 * param_3 := obj.Top();
 * param_4 := obj.GetMin();
 */
